### PR TITLE
REF: implement helpers for __array_ufunc__

### PR DESCRIPTION
**Pull request by jbrockmendel**
Refactored out of a branch that implements `Index.__array_ufunc__`

### CODE REVIEW

File: pandas/core/arraylike.py [hunk 1]
```diff
@@ -375,8 +389,66 @@ def reconstruct(result):
...
+        if len(out) == 1:
+            out = out[0]
+        else:
+            raise NotImplementedError
+
+    _assign_where(out, result, where)
+    return out
+
+
+def _assign_where(out, result, where) -> None:
```
Comment (jreback) [inline pandas/core/arraylike.py:446]: type in future

File: pandas/core/arraylike.py [hunk 2]
```diff
@@ -375,8 +389,66 @@ def reconstruct(result):
...
+
+
+def _dispatch_ufunc_with_out(self, ufunc: np.ufunc, method: str, *inputs, **kwargs):
+    """
+    If we have an `out` keyword, then call the ufunc without `out` and then
+    set the result into the given `out`.
+    """
+
+    # Note: we assume _standardize_out_kwarg has already been called.
+    out = kwargs.pop("out")
```
Comment (jreback) [inline pandas/core/arraylike.py:418]: prob need to type in future

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
assume these are going to be used in next pass

**Comment by jbrockmendel** [comment]
> assume these are going to be used in next pass

yep

**Pull request merged**